paddle/framework/framework.md

```diff
@@ -0,0 +1,205 @@
+# Design Doc: Framework
+
+This design is after the learning of other deep learning systems and numerous group discussions.
+
+## Operation and Operator
+
+At the core of a deep learning system is the neural network.  A neural network is a directed graph of operations, where each operation is an instance of an *operator*, which is a C++ class with a `Run` method derived from the base class `Operator`. 
+
+### `OperationProto` 
+
+Users are supposed to describe neural networks by calling Python.  functions known as *operation creators*, which, in turn call a C++ function `paddle::framework::CreateOperation`.  This C++ function eases the adding of more language bindings.
+
+Python and C++ have different function call syntax, so we define the parameter of `paddle::framework::CreateOperation` a protobuf message `OperationProto`.
+
+### `OperatorProto`
+
+We'd like to generate operator creators automatically from C++ code so to keep Python code always updated.  So we need to describe each C++ class in a protobuf message `OperatorProto`. We also need to fill in an `OperatorProto` message for each operator class and expose these messages to Python function `paddle.framework.create_operation_creators`.  We call this filling and exposing mechanism *operator registration*.
+
+
+## Operators, Layers, Variables, Scope
+
+### Operators as Functions
+
+An operator is intrinsically a function, which has inputs and outputs.  For example, the functional representation of the GEMM operator is
+
+```cpp
+gemm(X, W, scale, act=LeRU) {
+  unactivated = scale * X * W
+  if !act {
+    return unactivated
+  }
+  return unactivated, act(unactivated, cap)
+}
+```
+
+Note that operators might call other operators.  In above example, `gemm` calls `act`.
+
+### Gradient Operators
+
+Each operator has a corresponding gradient operator that defines the gradient computation.
+
+### Layers
+
+If we describe a neural network by calling operator creators directly, our code would be lengthy.  It is easier to use layers creators, which, in addition to calling operator creators for the computation, creates model parameters.
+
+For example, `paddle.layer.fc` and `paddle.layer.conv` both call `paddle.op.gemm`. They should also create and initialize `W`, the layer's parameter, and `act` and `scale`, the attributes.
+
+### Variables
+
+We prefer to represent inputs, outputs, and attributes in Variables, so an operation's output can be used to feed and to configure  other operations.
+
+Variables should be able to hold the following types of values:
+
+- int
+- bool
+- float
+- half
+- string
+- tensor
+- operator
+- scope
+
+Some Python API proposals:
+
+```python
+x = paddle.variable.new("x") # create a variable of not yet known type
```

Comment:
This APIs could be unified by one `paddle.variable.new`?

```python
x_var = paddle.variable.new("x", 10)

y = paddle.variable.Tensor()
y_var = paddle.variable.new("y", y)

z = paddle.variable.new("z", "A string")

no_name = paddle.variable.new(name=None, val=100)  # actually I think each variable should have a name. It might have some independent scope to hold them.
```